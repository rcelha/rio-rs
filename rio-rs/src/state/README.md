Module that provides object persistence

The `rio_rs::state` module has all the traits and structs one needs
to start saving objects from your cluster to persistent storage.

# State

Trait for implementing how to deserialize data in, and serialize data out of
a service (trait object).

Although you need to implement this trait for each state you have in your application,
the `ManagedState` derive implements it for you.

# State Loader

Any type that implements `rio_rs::state_provider::StateLoader`.

The `StateLoader` defines an interface to load state data from a backend.

# State Saver

Any type that implements `rio_rs::state_provider::StateSaver`.

It defines an interface to save data into a persistence backend.

# State Provider

Any type that is both a `State Saver` and a `State Loader`.

It provides functions to load and deserialize state, and serialize and save state.
