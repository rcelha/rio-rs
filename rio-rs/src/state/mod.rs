#![doc = include_str!("README.md")]

use std::ops::Deref;

use crate::errors::LoadStateError;
use crate::registry::IdentifiableType;
use crate::{ServiceObject, WithId};
use async_trait::async_trait;
use serde::de::DeserializeOwned;
use serde::Serialize;

#[cfg(feature = "local")]
pub mod local;

#[cfg(feature = "redis")]
pub mod redis;

// #[cfg(feature = "sql")]
// pub mod sql;

#[cfg(feature = "sqlite")]
pub mod sqlite;

#[cfg(feature = "postgres")]
pub mod postgres;

/// The `StateLoader` defines an interface to load serialized state from a source
///
/// **important** This trait is not responsible for serializing it back to its
/// original type
///
/// TODO use a reader type instead of String on load fn
#[async_trait]
pub trait StateLoader: Send + Sync {
    /// <div class="warning">
    /// TODO
    ///
    /// This here can't be used right now as it depends on the type `T` when
    /// called.
    ///
    /// This makes its usage quite cluncky (if not impossible) as you need to invoke it
    /// with a target of this loader
    /// </div>
    async fn prepare(&self) {}

    async fn load<T>(
        &self,
        object_kind: &str,
        object_id: &str,
        state_type: &str,
    ) -> Result<T, LoadStateError>
    where
        T: DeserializeOwned;
}

/// Auto implement [StateLoader] for every type that derefs to a [StateLoader]
///
/// This way you can create a wrapper for a [StateLoader] and it will automatically
/// get this implementation
#[async_trait]
impl<T, S> StateLoader for T
where
    T: Deref<Target = S> + Send + Sync,
    S: StateLoader,
{
    async fn load<O: DeserializeOwned>(
        &self,
        object_kind: &str,
        object_id: &str,
        state_type: &str,
    ) -> Result<O, LoadStateError> {
        self.deref().load(object_kind, object_id, state_type).await
    }
}

/// The `StateSave` defines an interface to save serialized data into a persistence
/// backend (memory, sql server, etc)
///
/// **important** This trait is not responsible for serializing the state from
/// its original type
#[async_trait]
pub trait StateSaver: Sync + Send {
    async fn prepare(&self) {}

    async fn save(
        &self,
        object_kind: &str,
        object_id: &str,
        state_type: &str,
        data: &(impl Serialize + Send + Sync),
    ) -> Result<(), LoadStateError>;
}

/// Auto implement [StateSaver] for every type that derefs to a StateSaver
///
/// This way you can create a wrapper for a [StateSaver] and it will automatically
/// get this implementation
#[async_trait]
impl<T, S> StateSaver for T
where
    T: Deref<Target = S> + Send + Sync,
    S: StateSaver,
{
    async fn save(
        &self,
        object_kind: &str,
        object_id: &str,
        state_type: &str,
        data: &(impl Serialize + Send + Sync),
    ) -> Result<(), LoadStateError> {
        self.deref()
            .save(object_kind, object_id, state_type, data)
            .await
    }
}

/// Reponsible for managing states for a specific object
///
/// With this trait one can load/save individual states from an orig (Self) object
#[async_trait]
pub trait ObjectStateManager {
    async fn load_state<T, S>(&mut self, state_loader: &S) -> Result<(), LoadStateError>
    where
        T: IdentifiableType + Serialize + DeserializeOwned + Default, // neends default cause of trait State
        S: StateLoader + Send + Sync,
        Self: State<T> + IdentifiableType + WithId + Send + Sync,
    {
        let object_kind = Self::user_defined_type_id();
        let object_id = self.id();
        let state_type = T::user_defined_type_id();
        let data: T = self
            .load(state_loader, object_kind, object_id, state_type)
            .await
            .or(Err(LoadStateError::ObjectNotFound))?;

        self.set_state(data);
        Ok(())
    }

    async fn save_state<T, S>(&self, state_saver: &S) -> Result<(), LoadStateError>
    where
        T: IdentifiableType + Serialize + DeserializeOwned + Sync + Default, // Needs default cause of trait State
        S: StateSaver,
        Self: State<T> + IdentifiableType + WithId + Send + Sync,
    {
        let object_kind = Self::user_defined_type_id();
        let object_id = self.id();

        let state_type = T::user_defined_type_id();
        let state_value: &T = self.get_state();
        state_saver
            .save(object_kind, object_id, state_type, &state_value)
            .await?;
        Ok(())
    }
}

// If an struct implements ServiceObject, it gets ObjectStateManager out of the box
impl<T> ObjectStateManager for T where T: ServiceObject {}

/// Trait to define how to get and set states in and out of an object
///
/// One need to implement this trait for each state a object holds
#[async_trait]
pub trait State<T>
where
    T: Serialize + DeserializeOwned + Default,
{
    fn get_state(&self) -> &T;
    fn set_state(&mut self, value: T);

    async fn load<S: StateLoader + Sync + Send>(
        &self,
        state_loader: &S,
        object_kind: &str,
        object_id: &str,
        state_type: &str,
    ) -> Result<T, LoadStateError> {
        state_loader.load(object_kind, object_id, state_type).await
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use rio_macros::{ManagedState, TypeName, WithId};
    use serde::Deserialize;

    type TestResult = Result<(), Box<dyn std::error::Error>>;

    #[derive(Default, Debug, PartialEq, Serialize, Deserialize)]
    struct PersonState {
        name: String,
        age: u8,
    }

    impl IdentifiableType for Option<PersonState> {
        fn instance_type_id(&self) -> &'static str {
            "OptionPersonState"
        }
    }

    #[derive(Default, Debug, Serialize, Deserialize, PartialEq, TypeName)]
    #[rio_path = "crate"]
    struct LegalPersonState {
        legal_name: String,
        id_document: String,
    }

    #[tokio::test]
    async fn sanity_check() -> TestResult {
        let local_state = local::LocalState::new();
        let state = PersonState {
            name: "Foo".to_string(),
            age: 21,
        };
        local_state.save("a", "1", "PersonState", &state).await?;
        let new_state: PersonState = local_state.load("a", "1", "PersonState").await?;
        assert_eq!(state, new_state);
        Ok(())
    }

    #[tokio::test]
    async fn model_call() -> TestResult {
        #[derive(Debug, Default, WithId, TypeName, ManagedState)]
        #[rio_path = "crate"]
        struct Person {
            id: String,
            #[managed_state]
            person_state: Option<PersonState>,
            #[managed_state]
            legal_state: LegalPersonState,
        }
        impl ObjectStateManager for Person {}

        impl Person {
            async fn load_all_states(
                &mut self,
                state_loader: &local::LocalState,
            ) -> Result<(), LoadStateError> {
                let _ = self
                    .load_state::<Option<PersonState>, _>(state_loader)
                    .await?;
                let _ = self.load_state::<LegalPersonState, _>(state_loader).await?;
                Ok(())
            }

            async fn save_all_states<S: StateSaver>(
                &mut self,
                state_saver: &S,
            ) -> Result<(), LoadStateError> {
                self.save_state::<Option<PersonState>, _>(state_saver)
                    .await?;
                self.save_state::<LegalPersonState, _>(state_saver).await?;
                Ok(())
            }
        }

        let local_state = local::LocalState::new();

        {
            let mut person = Person::default();
            person.person_state = Some(PersonState {
                name: "Foo".to_string(),
                age: 22,
            });
            person.legal_state = LegalPersonState {
                legal_name: "Foo Bla".to_string(),
                id_document: "123.123.123-12".to_string(),
            };
            person.save_all_states(&local_state).await?;
        }
        {
            let mut person = Person::default();
            person.load_all_states(&local_state).await?;
            assert!(person.person_state.is_some());
            assert_eq!(&person.legal_state.legal_name, "Foo Bla");
        }
        Ok(())
    }
}
