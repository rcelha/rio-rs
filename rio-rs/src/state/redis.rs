use async_trait::async_trait;
use bb8_redis::{bb8::Pool, RedisConnectionManager};
use redis::AsyncCommands;
use serde::de::DeserializeOwned;
use serde::Serialize;

use super::{StateLoader, StateSaver};
use crate::errors::LoadStateError;

/// State Storage using Redis/Valkey
#[derive(Clone)]
pub struct RedisState {
    pool: Pool<RedisConnectionManager>,
    key_prefix: String,
}

impl RedisState {
    pub async fn from_connect_string(connection_string: &str, key_prefix: Option<String>) -> Self {
        let conn_manager = RedisConnectionManager::new(connection_string).expect("TODO");
        let pool = Pool::builder().build(conn_manager).await.expect("TODO");
        let key_prefix = key_prefix.unwrap_or_default();
        Self { pool, key_prefix }
    }
}

#[async_trait]
impl<T: DeserializeOwned> StateLoader<T> for RedisState {
    async fn load(
        &self,
        object_kind: &str,
        object_id: &str,
        state_type: &str,
    ) -> Result<T, LoadStateError> {
        let object_kind = object_kind.to_string();
        let object_id = object_id.to_string();
        let state_type = state_type.to_string();
        let key = format!(
            "{}state:{}:{}:{}",
            self.key_prefix, object_kind, object_id, state_type
        );
        let mut client = self.pool.get().await.map_err(|_| LoadStateError::Unknown)?;
        let se_data: Option<String> = client.get(key).await.expect("TODO");
        if let Some(x) = se_data {
            let data = serde_json::from_str(&x);
            data.map_err(|_| LoadStateError::DeserializationError)
        } else {
            Err(LoadStateError::ObjectNotFound)
        }
    }
}

#[async_trait]
impl<T: Serialize + Send + Sync> StateSaver<T> for RedisState {
    async fn save(
        &self,
        object_kind: &str,
        object_id: &str,
        state_type: &str,
        data: &T,
    ) -> Result<(), LoadStateError> {
        let object_kind = object_kind.to_string();
        let object_id = object_id.to_string();
        let state_type = state_type.to_string();
        let key = format!(
            "{}state:{}:{}:{}",
            self.key_prefix, object_kind, object_id, state_type
        );
        let ser_data =
            serde_json::to_string(&data).map_err(|_| LoadStateError::SerializationError)?;
        let mut client = self.pool.get().await.map_err(|_| LoadStateError::Unknown)?;
        let _: () = client
            .set(key, ser_data)
            .await
            .map_err(|_| LoadStateError::Unknown)?;
        Ok(())
    }
}
