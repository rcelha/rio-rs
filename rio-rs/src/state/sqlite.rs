use crate::{errors::LoadStateError, sql_migration::SqlMigrations};
use async_trait::async_trait;
use futures::TryFutureExt;
use serde::{de::DeserializeOwned, Serialize};
use sqlx::{
    self,
    sqlite::{SqlitePoolOptions, SqliteRow},
    Row, SqlitePool,
};

use super::{StateLoader, StateSaver};

pub struct SqliteStateMigrations {}

impl SqlMigrations for SqliteStateMigrations {
    fn queries() -> Vec<String> {
        let migration_001 = include_str!("./migrations/0001-sqlite-init.sql");
        vec![migration_001.to_string()]
    }
}

#[derive(Debug)]
pub struct SqliteState {
    pool: SqlitePool,
}

impl SqliteState {
    pub fn pool() -> SqlitePoolOptions {
        SqlitePoolOptions::new()
    }

    pub fn new(pool: SqlitePool) -> Self {
        Self { pool }
    }

    pub async fn migrate(&self) {
        let mut transaction = self.pool.begin().await.unwrap();
        let queries = SqliteStateMigrations::queries();
        for query in queries {
            sqlx::query(&query).execute(&mut transaction).await.unwrap();
        }
        transaction.commit().await.unwrap();
    }
}

#[async_trait]
impl StateLoader for SqliteState {
    async fn prepare(&self) {
        self.migrate().await;
    }

    async fn load<T: DeserializeOwned>(
        &self,
        object_kind: &str,
        object_id: &str,
        state_type: &str,
    ) -> Result<T, LoadStateError> {
        let items = sqlx::query(
            r#"
            SELECT serialized_state
            FROM state_provider_object_state
            WHERE object_kind=$1 AND object_id=$2 AND state_type = $3
            "#,
        )
        .bind(object_kind)
        .bind(object_id)
        .bind(state_type)
        .map(|x: SqliteRow| -> String {
            let tmp = x.get::<Vec<u8>, _>("serialized_state");
            String::from_utf8(tmp).expect("TODO")
        })
        .fetch_one(&self.pool)
        .map_err(|_| LoadStateError::ObjectNotFound)
        .await?;
        let data = serde_json::from_str(&items).expect("TODO");
        Ok(data)
    }
}

#[async_trait]
impl StateSaver for SqliteState {
    async fn prepare(&self) {
        self.migrate().await;
    }

    async fn save(
        &self,
        object_kind: &str,
        object_id: &str,
        state_type: &str,
        data: &(impl Serialize + Send + Sync),
    ) -> Result<(), LoadStateError> {
        let serialized_data = serde_json::to_string(&data).expect("TODO");
        sqlx::query(
            r#"
            INSERT INTO
                state_provider_object_state(object_kind, object_id, state_type, serialized_state)
            VALUES ($1, $2, $3, $4)
            ON CONFLICT(object_kind, object_id, state_type) DO UPDATE SET serialized_state=$4
            "#,
        )
        .bind(object_kind)
        .bind(object_id)
        .bind(state_type)
        .bind(serialized_data.bytes().collect::<Vec<_>>())
        .execute(&self.pool)
        .map_err(|e| {
            eprintln!("{:?}", e);
            LoadStateError::Unknown
        })
        .await
        .map(|_| ())
    }
}
