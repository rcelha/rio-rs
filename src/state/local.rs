use super::{StateLoader, StateSaver};
use crate::errors::LoadStateError;
use async_trait::async_trait;
use dashmap::DashMap;
use serde::de::DeserializeOwned;
use serde::Serialize;

/// `LocalState` is a state provider for testing purposes
///
/// It stores all the serialized states into a single `DashMap`
#[derive(Debug)]
pub struct LocalState {
    data: DashMap<(String, String, String), String>,
}

impl LocalState {
    pub fn new() -> LocalState {
        LocalState {
            data: DashMap::new(),
        }
    }
}

#[async_trait]
impl StateLoader for LocalState {
    async fn load<T: DeserializeOwned>(
        &self,
        object_kind: &str,
        object_id: &str,
        state_type: &str,
    ) -> Result<T, LoadStateError> {
        let object_kind = object_kind.to_string();
        let object_id = object_id.to_string();
        let state_type = state_type.to_string();
        let k = (object_kind, object_id, state_type);

        if let Some(x) = self.data.get(&k) {
            Ok(serde_json::from_str(&x).expect("TODO"))
        } else {
            Err(LoadStateError::Unknown)
        }
    }
}

#[async_trait]
impl StateSaver for LocalState {
    async fn save(
        &self,
        object_kind: &str,
        object_id: &str,
        state_type: &str,
        data: &(impl Serialize + Send + Sync),
    ) -> Result<(), LoadStateError> {
        let object_kind = object_kind.to_string();
        let object_id = object_id.to_string();
        let state_type = state_type.to_string();
        let k = (object_kind, object_id, state_type);
        self.data
            .insert(k, serde_json::to_string(&data).expect("TODO"));
        Ok(())
    }
}

#[cfg(test)]
mod test {
    use rio_macros::TypeName;
    use serde::Deserialize;

    use super::*;

    type TestResult = Result<(), Box<dyn std::error::Error>>;

    #[derive(TypeName, Debug, Serialize, Deserialize, PartialEq)]
    #[rio_path = "crate"]
    struct TestState {
        name: String,
    }

    #[tokio::test]
    async fn sanity_check() -> TestResult {
        let local_state = LocalState::new();
        let state = TestState {
            name: "Foo".to_string(),
        };
        local_state.save("a", "1", "TestState", &state).await?;
        let new_state: TestState = local_state.load("a", "1", "TestState").await?;
        assert_eq!(state, new_state);
        Ok(())
    }
}
