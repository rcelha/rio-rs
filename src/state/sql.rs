use crate::errors::LoadStateError;
use async_trait::async_trait;
use futures::TryFutureExt;
use serde::{de::DeserializeOwned, Serialize};
use sqlx::{
    self,
    any::{AnyPoolOptions, AnyRow},
    AnyPool, Row,
};

use super::{StateLoader, StateSaver};

#[derive(Debug)]
pub struct SqlState {
    pool: AnyPool,
}

impl SqlState {
    pub fn pool() -> AnyPoolOptions {
        AnyPoolOptions::new()
    }

    pub fn new(pool: AnyPool) -> Self {
        Self { pool }
    }

    pub async fn migrate(&self) {
        let mut transaction = self.pool.begin().await.unwrap();

        let queries = [
            // Table with all state objects
            r#"CREATE TABLE IF NOT EXISTS state_provider_object_state
               (
                   object_kind       TEXT                              NOT NULL,
                   object_id         TEXT                              NOT NULL,
                   state_type       TEXT                              NOT NULL,
                   serialized_state BLOB                              NOT NULL,
                   PRIMARY KEY (object_kind, object_id, state_type)
               )"#,
        ];
        for query in queries {
            sqlx::query(query).execute(&mut transaction).await.unwrap();
        }
        transaction.commit().await.unwrap();
    }
}

#[async_trait]
impl StateLoader for SqlState {
    async fn load<T: DeserializeOwned>(
        &self,
        object_kind: &str,
        object_id: &str,
        state_type: &str,
    ) -> Result<T, LoadStateError> {
        let items = sqlx::query(
            r#"
            SELECT serialized_state
            FROM state_provider_object_state
            WHERE object_kind=$1 AND object_id=$2 AND state_type = $3
            "#,
        )
        .bind(object_kind)
        .bind(object_id)
        .bind(state_type)
        .map(|x: AnyRow| -> String { x.get("serialized_state") })
        .fetch_one(&self.pool)
        .map_err(|_| LoadStateError::ObjectNotFound)
        .await?;
        let data = serde_json::from_str(&items).expect("TODO");
        Ok(data)
    }
}

#[async_trait]
impl StateSaver for SqlState {
    async fn save(
        &self,
        object_kind: &str,
        object_id: &str,
        state_type: &str,
        data: &(impl Serialize + Send + Sync),
    ) -> Result<(), LoadStateError> {
        let serialized_data = serde_json::to_string(&data).expect("TODO");
        sqlx::query(
            r#"
            INSERT INTO
                state_provider_object_state(object_kind, object_id, state_type, serialized_state)
            VALUES ($1, $2, $3, $4)
            ON CONFLICT(object_kind, object_id, state_type) DO UPDATE SET serialized_state=$4
            "#,
        )
        .bind(object_kind)
        .bind(object_id)
        .bind(state_type)
        .bind(serialized_data)
        .execute(&self.pool)
        .map_err(|_| LoadStateError::Unknown)
        .await
        .map(|_| ())
    }
}
