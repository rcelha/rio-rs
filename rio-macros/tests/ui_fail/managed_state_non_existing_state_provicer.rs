use rio_macros::*;
use rio_rs::state::ObjectStateManager;
use rio_rs::ServiceObject;
use serde::{Deserialize, Serialize};

#[derive(TypeName, Serialize, Deserialize, Default)]
struct TestVec(Vec<usize>);

#[derive(Default, WithId, TypeName, ManagedState)]
struct TestService {
    id: String,
    // The provider below is not imported
    #[managed_state(provider = SqliteState)]
    tests: TestVec,
}

impl ServiceObject for TestService {}

fn main() {}
