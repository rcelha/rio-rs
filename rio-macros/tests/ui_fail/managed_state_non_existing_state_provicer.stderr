error[E0412]: cannot find type `SqliteState` in this scope
  --> tests/ui_fail/managed_state_non_existing_state_provicer.rs:13:32
   |
13 |     #[managed_state(provider = SqliteState)]
   |                                ^^^^^^^^^^^ not found in this scope
   |
help: consider importing this struct
   |
1  + use rio_rs::state::sqlite::SqliteState;
   |
